File: Answer/1722.执行交换操作后的最小汉明距离.go
/*
 * @lc app=leetcode.cn id=1722 lang=golang
 *
 * [1722] 执行交换操作后的最小汉明距离
 */

// @lc code=start
func minimumHammingDistance(source []int, target []int, allowedSwaps [][]int) int {
	n := len(source)
	parent := make([]int, n)
	sz := make([]int, n)
	for i := range parent {
		parent[i] = i
		sz[i] = 1
	}

	// 迭代路径压缩，避免递归开销
	find := func(x int) int {
		root := x
		for parent[root] != root {
			root = parent[root]
		}
		for parent[x] != x {
			parent[x], x = root, parent[x]
		}
		return root
	}

	// 按大小合并，保持树平衡
	for _, swap := range allowedSwaps {
		a, b := swap[0], swap[1]
		pa, pb := find(a), find(b)
		if pa != pb {
			if sz[pa] < sz[pb] {
				pa, pb = pb, pa
			}
			parent[pb] = pa
			sz[pa] += sz[pb]
		}
	}

	// 缓存每个位置的最终根节点，避免后续重复查找
	roots := make([]int, n)
	for i := 0; i < n; i++ {
		roots[i] = find(i)
	}

	// 按连通分量统计 source 中各数值出现次数
	groups := make(map[int]map[int]int, n)
	for i := 0; i < n; i++ {
		r := roots[i]
		if groups[r] == nil {
			groups[r] = make(map[int]int)
		}
		groups[r][source[i]]++
	}

	res := 0
	for i := 0; i < n; i++ {
		r := roots[i]
		if groups[r][target[i]] > 0 {
			groups[r][target[i]]--
		} else {
			res++
		}
	}
	return res
}

// @lc code=end

File: Answer/1722.执行交换操作后的最小汉明距离.java
/*
 * @lc app=leetcode.cn id=1722 lang=java
 *
 * [1722] 执行交换操作后的最小汉明距离
 */

// @lc code=start
class Solution {
    public int minimumHammingDistance(int[] source, int[] target, int[][] allowedSwaps) {
        int n = source.length;
        int[] parent = new int[n];
        int[] sz = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            sz[i] = 1;
        }

        for (int[] swap : allowedSwaps) {
            int pa = find(parent, swap[0]);
            int pb = find(parent, swap[1]);
            if (pa != pb) {
                if (sz[pa] < sz[pb]) {
                    int tmp = pa; pa = pb; pb = tmp;
                }
                parent[pb] = pa;
                sz[pa] += sz[pb];
            }
        }

        int[] roots = new int[n];
        for (int i = 0; i < n; i++) {
            roots[i] = find(parent, i);
        }

        java.util.HashMap<Integer, java.util.HashMap<Integer, Integer>> groups = new java.util.HashMap<>();
        for (int i = 0; i < n; i++) {
            int r = roots[i];
            groups.computeIfAbsent(r, k -> new java.util.HashMap<>()).merge(source[i], 1, Integer::sum);
        }

        int res = 0;
        for (int i = 0; i < n; i++) {
            int r = roots[i];
            var cnt = groups.get(r);
            if (cnt.getOrDefault(target[i], 0) > 0) {
                cnt.put(target[i], cnt.get(target[i]) - 1);
            } else {
                res++;
            }
        }
        return res;
    }

    private int find(int[] parent, int x) {
        int root = x;
        while (parent[root] != root) {
            root = parent[root];
        }
        while (parent[x] != x) {
            int next = parent[x];
            parent[x] = root;
            x = next;
        }
        return root;
    }
}
// @lc code=end
